src/RateLimiter/Counter.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\Http\Message\ServerRequestInterface;
+
+final class Counter
+{
+    private int $interval = 360;
+
+    private ?string $id = null;
+
+    /**
+     * @var callable
+     */
+    private $idCallback;
+
+    private bool $init = false;
+
+    private CounterStorageInterface $storage;
+
+    public function __construct(CounterStorageInterface $storage)
+    {
+        $this->storage = $storage;
+    }
+
+    public function init(ServerRequestInterface $request): self
+    {
+        $this->id = $this->generateId($request);
+
+        if (!$this->storage->has($this->id)) {
+            $this->storage->set($this->id, 0, $this->interval);
+        }
+
+        $this->init = true;
+
+        return $this;
+    }
+
+    public function setIdCallback(callable $callback): self
+    {
+        $this->idCallback = $callback;
+
+        return $this;
+    }
+
+    public function setId(string $id): self
+    {
+        $this->id = $id;
+
+        return $this;
+    }
+
+    public function setInterval(int $interval): self
+    {
+        $this->interval = $interval;
+
+        return $this;
+    }
+
+    public function increment(): void
+    {
+        $value = $this->getCounterValue();
+        $value++;
+
+        $this->storage->set($this->id, $value, $this->interval);
+    }
+
+    public function getCounterValue(): int
+    {
+        $this->checkInit();
+        return $this->storage->get($this->id);
+    }
+
+    private function hasIdCallback(): bool
+    {
+        return $this->idCallback !== null;
+    }
+
+    private function generateId(ServerRequestInterface $request): string
+    {
+        if ($this->hasIdCallback()) {
+            return \call_user_func($this->idCallback, $request);
+        }
+
+        return $this->id ?? $this->generateIdFromRequest($request);
+    }
+
+    private function generateIdFromRequest(ServerRequestInterface $request): string
+    {
+        return strtolower('rate-limiter-' . $request->getMethod() . '-' . $request->getUri()->getPath());
```

Comment:
generate is not public, the `strtolower()` can be omitted.

Reply:
Query request parameters can be anything, I just want the ID to be the same in terms of case.